core/src/main/java/org/apache/calcite/rel/rules/SortRemoveRule.java

```diff
@@ -63,5 +83,27 @@ public SortRemoveRule(RelBuilderFactory relBuilderFactory) {
         .getTrait(RelCollationTraitDef.INSTANCE);
     final RelTraitSet traits = sort.getInput().getTraitSet().replace(collation);
     call.transformTo(convert(sort.getInput(), traits));
+    return true;
+  }
+
+  // Sort is removed if all of following conditions are met
+  // 1. Sort's input is guaranteed to produce at most one row
+  // 2. If fetch is greater than 0
+  // 3. If offset is less than 1
+  public static boolean shouldRemoveSortBasedOnRowCount(final RelMetadataQuery mq,
+      final RelNode input, int fetch, int offset) {
+    Double maxRowCount = mq.getMaxRowCount(input);
+    if (maxRowCount != null && maxRowCount <= 1) {
+      if (fetch == 0) {
+        return false;
+      } else if (fetch > 0) {
+        if (offset < 1) {
+          return true;
+        }
+      } else {
+        return true;
```

Comment:
What if `fetch = -1` and `offset > 1`, should it return `true` ? If no, do we have test to cover ?

Reply:
I see that other databases like MySql, Hive etc do not allow OFFSET without LIMIT. But looks like Calcite does. I will add update the code to handle this case. Thanks @jinxing64